libbeat/processors/rate_limit/algorithm/token_bucket.go

```diff
@@ -0,0 +1,192 @@
+	BurstMultiplier float64 `config:"burst_multiplier"`
+
+	// GC governs when completely filled token buckets must be deleted
+	// to free up memory. GC is performed when _any_ of the GC conditions
+	// below are met. After each GC, counters corresponding to _each_ of
+	// the GC conditions below are reset.
+	GC tokenBucketGCConfig `config:"gc"`
+}
+
+func newTokenBucket(config Config) (Algorithm, error) {
+	cfg := tokenBucketConfig{
+		BurstMultiplier: 1.0,
+		GC: tokenBucketGCConfig{
+			NumCalls:   10000,
+			NumBuckets: 1000,
+		},
+	}
+
+	if err := config.Config.Unpack(&cfg); err != nil {
+		return nil, errors.Wrap(err, "could not unpack token_bucket algorithm configuration")
+	}
+
+	return &tokenBucket{
+		limit:   config.Limit,
+		depth:   config.Limit.value * cfg.BurstMultiplier,
+		buckets: sync.Map{},
+		gc: struct {
+			thresholds tokenBucketGCConfig
+			metrics    struct {
+				numCalls   atomic.Uint
+				numBuckets atomic.Uint
+			}
+		}{
+			thresholds: tokenBucketGCConfig{
+				NumCalls:   cfg.GC.NumCalls,
+				NumBuckets: cfg.GC.NumBuckets,
+			},
+		},
+		clock: clockwork.NewRealClock(),
+	}, nil
+}
+
+func (t *tokenBucket) IsAllowed(key uint64) bool {
+	t.runGC()
+
+	b := t.getBucket(key)
+	allowed := b.withdraw()
+
+	t.gc.metrics.numCalls.Inc()
+	return allowed
+}
+
+// SetClock allows test code to inject a fake clock
+func (t *tokenBucket) SetClock(c clockwork.Clock) {
+	t.clock = c
+}
+
+func (t *tokenBucket) getBucket(key uint64) *bucket {
+	v, exists := t.buckets.LoadOrStore(key, &bucket{
+		tokens:        t.depth,
+		lastReplenish: t.clock.Now(),
+	})
+	b := v.(*bucket)
+
+	if exists {
+		b.replenish(t.limit, t.clock)
+		return b
+	}
+
+	t.gc.metrics.numBuckets.Inc()
+	return b
+}
+
+func (b *bucket) withdraw() bool {
+	if b.tokens < 1 {
+		return false
+	}
+	b.tokens--
+	return true
+}
+
+func (b *bucket) replenish(rate Rate, clock clockwork.Clock) {
+	secsSinceLastReplenish := clock.Now().Sub(b.lastReplenish).Seconds()
+	tokensToReplenish := secsSinceLastReplenish * rate.valuePerSecond()
+
+	b.tokens += tokensToReplenish
+	b.lastReplenish = clock.Now()
+}
+
+func (t *tokenBucket) runGC() {
+	// Don't run GC if thresholds haven't been crossed.
+	if (t.gc.metrics.numBuckets.Load() < t.gc.thresholds.NumBuckets) &&
+		(t.gc.metrics.numCalls.Load() < t.gc.thresholds.NumCalls) {
+		return
+	}
+
+	// Add tokens to all buckets according to the rate limit
+	// and flag full buckets for deletion.
+	toDelete := make([]uint64, 0)
+	var numBuckets uint
+	t.buckets.Range(func(k, v interface{}) bool {
+		key := k.(uint64)
+		b := v.(*bucket)
+
+		b.replenish(t.limit, t.clock)
+
+		if b.tokens >= t.depth {
+			toDelete = append(toDelete, key)
+		}
+
+		numBuckets++
+
+		return true
+	})
+
+	// Cleanup full buckets to free up memory
+	for _, key := range toDelete {
+		t.buckets.Delete(key)
+		numBuckets--
+	}
+
+	// Reset GC metrics
+	t.gc.metrics.numCalls = atomic.MakeUint(0)
+	t.gc.metrics.numBuckets = atomic.MakeUint(numBuckets)
```

Comment:
If buckets are created while the GC is deleting buckets they will not be counted by `numBuckets`, and this will get out of sync. I think you can substract here the number of deleted ones:
```suggestion
	t.gc.metrics.numBuckets.Sub(len(toDelete))
```
And then `numBuckets` is not needed.

Reply:
Hmm, this is bringing up another couple of points:
- What if new buckets are being added to the map while GC is going on, specifically while the `t.buckets.Range(...)` loop is being executed. Will these new buckets be considered in that loop? Reading https://golang.org/pkg/sync/#Map.Range, my understanding is that in the worst case some new buckets will be missed in this GC cycle. And that should be okay IMO as they would be considered in the next one.
- Consider a bucket that is full during the `t.buckets.Range(...)` loop and therefore gets flagged for deletion (added to `toDelete`). But before the `for _, key := range toDelete` loop gets executed, another goroutine withdraws tokens from the same bucket, therefore making it no longer a candidate for deletion. I think this is a dangerous case as it affects the correctness of the algorithm.

Given the above, I'm wondering now if token withdrawal (specifically the subtraction part) should also try to acquire the same lock as GC? That way, while GC is ongoing, token withdrawal is waiting on the lock, thereby allowing the entire GC operation to operate on a consistent snapshot of the buckets. WDYT?

Reply:
> What if new buckets are being added to the map while GC is going on, specifically while the t.buckets.Range(...) loop is being executed. Will these new buckets be considered in that loop? Reading https://golang.org/pkg/sync/#Map.Range, my understanding is that in the worst case some new buckets will be missed in this GC cycle. And that should be okay IMO as they would be considered in the next one.

Agree, this wouldn't be a problem, it is probably ok to not remove newer buckets.

> Consider a bucket that is full during the t.buckets.Range(...) loop and therefore gets flagged for deletion (added to toDelete). But before the for _, key := range toDelete loop gets executed, another goroutine withdraws tokens from the same bucket, therefore making it no longer a candidate for deletion. I think this is a dangerous case as it affects the correctness of the algorithm.

Not sure if this is such a problem, this will mean that producers that produce events on bursts (giving time between bursts to re-fill their buckets) will sometimes behave as if they have slightly bigger buckets. Producers that exceed its quota on a constant rate will still be correctly limited.

Something that can be a problem, related to this, is that if there are more buckets in use than the buckets threshold, the GC is going to be triggered constantly, so any full bucket will have more chances to be deleted when it is in use.
The number of buckets is probably going to be quite constant in a normal deployment, so maybe this threshold can be removed: in deployments with few buckets, this is never going to trigger a GC, in deployments with many buckets, it may be constantly triggering the GC. If not removed, maybe this should be a dynamic threshold, and behave in a similar way to the Go GC regarding the `GOGC` value, but I'd go for simplicity by now, and probably triggering the GC every n calls is enough.

> Given the above, I'm wondering now if token withdrawal (specifically the subtraction step) should also try to acquire the same lock as GC? That way, while GC is ongoing, token withdrawal is waiting on the lock, thereby allowing the entire GC operation to operate on a consistent snapshot of the buckets. WDYT?

A goroutine in the withdrawal step already has a reference to the bucket, so it will still use the old bucket, that is going to be deleted if marked by deletion in the `r.buckets.Range()`. The lock would need to go before `getBucket`, so no goroutine can get a bucket that is in the process of being deleted. But if we add a lock here (or in `withdrawal`) maybe we are adding too much contention.

Reply:
> Not sure if this is such a problem, this will mean that producers that produce events on bursts (giving time between bursts to re-fill their buckets) will sometimes behave as if they have slightly bigger buckets. Producers that exceed its quota on a constant rate will still be correctly limited.

Hmmm, I suppose there is some pathological case where the GC is happening at exactly the wrong time because there is a bucket that is always being flagged for deletion in the GC'ing goroutine, getting withdrawn from in another goroutine, and then being deleted back in the GC'ing goroutine. The impact would be, as you pointed out, that this particular bucket behaves as if it's slightly larger than it should be. 

But I suppose we are relying on chance and imperfect event rates to prevent such a situation from consistently occuring. I'm wondering if, in addition to the configurable threshold(s), we should introduce some chance as well, when deciding whether to run GC, e.g. roll a 6-sided dice and if the number is < 3, don't run GC that time?

> Something that can be a problem, related to this, is that if there are more buckets in use than the buckets threshold, the GC is going to be triggered constantly, so any full bucket will have more chances to be deleted when it is in use.
The number of buckets is probably going to be quite constant in a normal deployment, so maybe this threshold can be removed: in deployments with few buckets, this is never going to trigger a GC, in deployments with many buckets, it may be constantly triggering the GC.

Yeah, thinking about it some more, I agree it's best to remove the `NumBuckets` threshold. In deployments with few buckets we're anyway going to end up relying on the `NumCalls` threshold for GC to be called and in deployments with many buckets we'd end up raising the `NumBuckets` threshold to high enough that we are then in the same scenario as the "few buckets" deployment). Indeed finding a decent `NumBuckets` threshold that sometimes gets crossed but not all the time may require a fair bit of trial and error. Instead, we could just simplify and remove this threshold altogether.

Related to this discussion, I'm wondering if we want to emit a debug-level log when GC is triggered with a bit of info of the # of buckets in the map? It might be helpful to determine if GC is happening too frequently or not frequently enough and tune the `NumCalls` threshold accordingly?

> A goroutine in the withdrawal step already has a reference to the bucket, so it will still use the old bucket, that is going to be deleted if marked by deletion in the r.buckets.Range(). The lock would need to go before getBucket, so no goroutine can get a bucket that is in the process of being deleted. But if we add a lock here (or in withdrawal) maybe we are adding too much contention.

Well, my concern was more around the fact that we'd be deleting buckets that shouldn't be deleted (because between the step where the bucket is flagged for deletion and the step where the bucket is deleted from the map, there is a withdrawal from the bucket). But as you pointed out earlier, this just means that such buckets will behave as they are slightly bigger. So I think this point of locking on the same mutex in the withdrawal step, or even better before `getBucket` as you note, is moot now and we can ignore it.

Reply:
> But I suppose we are relying on chance and imperfect event rates to prevent such a situation from consistently occuring. I'm wondering if, in addition to the configurable threshold(s), we should introduce some chance as well, when deciding whether to run GC, e.g. roll a 6-sided dice and if the number is < 3, don't run GC that time?

I think that the lesser the options the better, we can trust by now on the imperfection of event rates to have a good enough randomization :slightly_smiling_face: We can add something later if we see this happening too frequently.

> Related to this discussion, I'm wondering if we want to emit a debug-level log when GC is triggered with a bit of info of the # of buckets in the map? It might be helpful to determine if GC is happening too frequently or not frequently enough and tune the NumCalls threshold accordingly?

+1 to log something at the debug level when a GC is triggered, this could contain apart of the number of buckets, the number of deleted buckets, and the duration of the process.

Reply:
> I think that the lesser the options the better, ...

Just to clarify, I wasn't suggesting that we add a user-configurable option for the randomization but just doing it as part of the implementation. Still, I'm okay with not doing even this much right now if you prefer.

Reply:
Removed `NumBuckets` threshold in 0d491c1c2b30a1a6d90a0ef3c92993b4a6761cea.
Implemented debug-level logging in 5c595b14934cbc45a03f08308430519eca82a000.